src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeTypeHandlerFactory.cs

```diff
@@ -0,0 +1,21 @@
+using System;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeHandling;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    class MappedCompositeTypeHandlerFactory<T> : NpgsqlTypeHandlerFactory<T>, IMappedCompositeTypeHandlerFactory
+        where T : new()
+    {
+        public INpgsqlNameTranslator NameTranslator { get; }
+
+        internal MappedCompositeTypeHandlerFactory(INpgsqlNameTranslator nameTranslator)
+            => NameTranslator = nameTranslator;
+
+        internal override NpgsqlTypeHandler Create(PostgresType pgType, NpgsqlConnection conn)
+            => MappedCompositeHandler<T>.Create((PostgresCompositeType)pgType, conn.Connector.TypeMapper, NameTranslator);
+
+        protected override NpgsqlTypeHandler<T> Create(NpgsqlConnection conn)
+            => throw new InvalidOperationException($"Expect {nameof(PostgresType)}");
```

Comment:
Should this be a `NotSupportedException`?

Reply:
Not mine. Just moved the old code to a separate file.

By the way, I think that this method can be removed from all factories, and the behavior should be changed to provide a Postgres type directly to the constructor. This means that the internal method must be `protected` not `internal`.

Reply:
The point here was that handler in general don't care about the PostgresType at all - except for CompositeTypeHandler. At the same time, they need to reference it for the metadata APIs (e.g. `NpgsqlDataReader.GetDataTypeName()`.

So we could add a constructor to each and every handler as @YohDeadfall suggests, and pass it up to the base class's constructor, or we can have it the current way where a type handler doesn't need to care about the PostgreSQL type, unless they choose to (like the composite handler). I remember not being too sure what I preferred, by the way, but I ended up doing this.

The specific exception type doesn't really matter since the method is never supposed to get invoked.